nncf/common/graph/graph.py

```diff
@@ -148,17 +148,23 @@ def __init__(
         self.parallel_input_port_ids = parallel_input_port_ids
 
     def __str__(self):
-        return str(self.from_node) + " -> " + str(self.tensor_shape) + " -> " + str(self.to_node)
+        return f"{self.from_node}:{self.output_port_id} -> {self.tensor_shape} -> {self.to_node}:{self.input_port_id}"
 
     def __hash__(self):
-        return hash(str(self))
+        return hash(
+            (
+                self.from_node,
+                self.to_node,
+                self.input_port_id,
+                self.output_port_id,
+                tuple(self.tensor_shape),
+                self.dtype,
+                tuple(self.parallel_input_port_ids),
+            )
+        )
```

Comment:
Are all these properties required for unique identification of an edge? I'm not sure how frequently edges are used as dict keys, but such a big list of things to be hashed may worsen the performance.

Reply:
That could affect performance. I think the best solution will be dataclass